File: db4o.net/Db4oUnit/Db4oUnit/Db4oUnit/Mocking/MethodCall.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4oUnit;
using Db4oUnit.Mocking;
using Db4objects.Db4o.Foundation;

namespace Db4oUnit.Mocking
{
	public class MethodCall
	{
		private sealed class _object_9 : object
		{
			public _object_9()
			{
			}

			public override string ToString()
			{
				return "...";
			}
		}

		public static readonly object IgnoredArgument = new _object_9();

		public interface IArgumentCondition
		{
			void Verify(object argument);
		}

		public class Conditions
		{
			public static MethodCall.IArgumentCondition IsA(Type expectedClass)
			{
				return new _IArgumentCondition_21(expectedClass);
			}

			private sealed class _IArgumentCondition_21 : MethodCall.IArgumentCondition
			{
				public _IArgumentCondition_21(Type expectedClass)
				{
					this.expectedClass = expectedClass;
				}

				public void Verify(object argument)
				{
					Assert.IsInstanceOf(expectedClass, argument);
				}

				private readonly Type expectedClass;
			}
		}

		public readonly string methodName;

		public readonly object[] args;

		public MethodCall(string methodName, object[] args)
		{
			this.methodName = methodName;
			this.args = args;
		}

		public override string ToString()
		{
			return methodName + "(" + Iterators.Join(Iterators.Iterate(args), ", ") + ")";
		}

		public override bool Equals(object obj)
		{
			if (null == obj)
			{
				return false;
			}
			if (GetType() != obj.GetType())
			{
				return false;
			}
			MethodCall other = (MethodCall)obj;
			if (!methodName.Equals(other.methodName))
			{
				return false;
			}
			if (args.Length != other.args.Length)
			{
				return false;
			}
			for (int i = 0; i < args.Length; ++i)
			{
				object expectedArg = args[i];
				if (expectedArg == IgnoredArgument)
				{
					continue;
				}
				object actualArg = other.args[i];
				if (expectedArg is MethodCall.IArgumentCondition)
				{
					((MethodCall.IArgumentCondition)expectedArg).Verify(actualArg);
					continue;
				}
				if (!Check.ObjectsAreEqual(expectedArg, actualArg))
				{
					return false;
				}
			}
			return true;
		}
	}
}

File: db4ounit/src/db4ounit/mocking/MethodCall.java
package db4ounit.mocking;

import com.db4o.foundation.*;

import db4ounit.*;

public class MethodCall {
	
	public static final Object IGNORED_ARGUMENT = new Object() {
		public String toString() {
			return "...";
		}
	};
	
	public static interface ArgumentCondition {
		void verify(Object argument);
	}
	
	public static class Conditions {
		public static ArgumentCondition isA(final Class<?> expectedClass) {
			return new ArgumentCondition() {
				public void verify(Object argument) {
					Assert.isInstanceOf(expectedClass, argument);
                }
			};
		}
	}
	
	public final String methodName;
	public final Object[] args;
	
	public MethodCall(String methodName, Object... args) {
		this.methodName = methodName;
		this.args = args;
	}

	public String toString() {
		return methodName + "(" + Iterators.join(Iterators.iterate(args), ", ") + ")";
	}
	
	public boolean equals(Object obj) {
		if (null == obj) return false;
		if (getClass() != obj.getClass()) return false;
		MethodCall other = (MethodCall)obj;
		if (!methodName.equals(other.methodName)) return false;
		if (args.length != other.args.length) return false;
		for (int i=0; i<args.length; ++i) {
			final Object expectedArg = args[i];
			if (expectedArg == IGNORED_ARGUMENT) {
				continue;
			}
			final Object actualArg = other.args[i];
			if (expectedArg instanceof ArgumentCondition) {
				((ArgumentCondition)expectedArg).verify(actualArg);
				continue;
			}
			
			if (!Check.objectsAreEqual(expectedArg, actualArg)) {
				return false;
			}
		}
		return true;
	}
}
